{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import hydra\n",
    "import torch\n",
    "import itertools\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.decomposition import PCA\n",
    "from omegaconf import OmegaConf\n",
    "from pytorch_lightning import (\n",
    "    LightningDataModule,\n",
    "    LightningModule,\n",
    "    seed_everything,\n",
    ")\n",
    "\n",
    "from multimodal_contrastive.analysis.utils import *\n",
    "from multimodal_contrastive.utils import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme(style=\"whitegrid\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# register custom resolvers if not already registered\n",
    "OmegaConf.register_new_resolver(\"sum\", lambda input_list: np.sum(input_list), replace=True)\n",
    "\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "device"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load config for CLIP model\n",
    "config_name = \"puma_sm_gmc\"\n",
    "configs_path = \"../../configs\"\n",
    "\n",
    "with hydra.initialize(version_base=None, config_path=configs_path):\n",
    "    cfg = hydra.compose(config_name=config_name)\n",
    "\n",
    "cfg.datamodule.split_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set seed for random number generators in pytorch, numpy and python.random\n",
    "# and especially for generating the same data splits for the test set\n",
    "if cfg.get(\"seed\"):\n",
    "    seed_everything(cfg.seed, workers=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load test data split\n",
    "datamodule: LightningDataModule = hydra.utils.instantiate(cfg.datamodule)\n",
    "datamodule.setup(\"test\")\n",
    "\n",
    "# train_loader = datamodule.train_dataloader()\n",
    "test_loader = datamodule.infer_dataloader()\n",
    "val_loader = datamodule.val_dataloader()\n",
    "train_idx, val_idx, test_idx = datamodule.get_split_idx()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load model from checkpoint\n",
    "ckpt_path = \"path.to/gmc_proxy.ckpt\"\n",
    "model: LightningModule = utils.instantiate_model(cfg)\n",
    "model = model.load_from_checkpoint(ckpt_path, map_location=device)\n",
    "model = model.eval()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set which loader to use\n",
    "# loader = test_loader\n",
    "loader, filtered_ge_idx = test_loader, test_idx\n",
    "loader = make_eval_data_loader(loader.dataset)\n",
    "\n",
    "# Peek at the first example in the loaded set\n",
    "example = next(iter(loader))\n",
    "inputs = example[\"inputs\"]\n",
    "struct, morph = inputs[\"struct\"], inputs[\"morph\"]\n",
    "\n",
    "print(struct)\n",
    "print(morph.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get latent representations for all test examples\n",
    "representations, mols = model.compute_representation_dataloader(\n",
    "    loader,\n",
    "    device=device,\n",
    "    return_mol=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If dataset is too large, randomly subset to cutoff molecules\n",
    "p_mols, p_representations, p_idx = random_subset(mols, representations, cutoff=3000)\n",
    "p_representations[\"morph\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use rdkit to get Morgan fingerprints for each molecule in mols\n",
    "mol_fps = get_molecular_fingerprints(p_mols, fp_type=\"morgan\", nbits=2048)\n",
    "tanimoto_sim = get_molecular_similarity(mol_fps, \"tanimoto\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the raw domain data (gene expression & cell profiler features) from the dataloader\n",
    "mods = unroll_dataloader(loader, mods=['morph'])\n",
    "# ge = mods['ge'][p_idx]\n",
    "morph = mods['morph'][p_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run pca on raw gene expression matrix\n",
    "N_genes=30\n",
    "pca = PCA(n_components=N_genes)\n",
    "ge_pca = pca.fit_transform(ge)\n",
    "ge_pca.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run pca on raw morphology (Cell Profiler) features\n",
    "N_morph_f = 30\n",
    "pca = PCA(n_components=N_morph_f)\n",
    "morph_pca = pca.fit_transform(morph)\n",
    "morph_pca.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset raw gene expression to top-k=81 most differentially expressed genes\n",
    "filtered_ge = np.load(\"adata_scale_filt.npy\")\n",
    "filtered_ge = filtered_ge[filtered_ge_idx][p_idx]\n",
    "filtered_ge.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Single modality latent distance versus tanimoto similarity\n",
    "mods_x = ['struct', 'morph', 'joint']\n",
    "mods_y = tanimoto_sim\n",
    "\n",
    "fig, ax = plt.subplots(2, 3, figsize=(15, 10))\n",
    "\n",
    "for idx, mod_x in enumerate(mods_x):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_x],\n",
    "        p_representations[mod_x],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    x, y = get_values_from_dist_mat(x, mods_y, keep_diag=False)\n",
    "    corr = np.corrcoef(x, y)\n",
    "    mod_x_label = mod_x.capitalize()\n",
    "\n",
    "    # Sample a subset of the data for visualization\n",
    "    rnd_idx = np.random.choice(len(x), 15000, replace=False)\n",
    "    x, y = x[rnd_idx], y[rnd_idx]\n",
    "\n",
    "    ax[0, idx].scatter(x, y, alpha=1, s=1)\n",
    "    ax[0, idx].set_xlabel(f\"Cosine Similarity ({mod_x}-{mod_x})\")\n",
    "    ax[0, idx].set_ylabel(\"Tanimoto Similarity\")\n",
    "    ax[0, idx].set_title(f\"Latent {mod_x_label} Cosine Sim. vs Tanimoto Sim.\")\n",
    "    ax[0, idx].legend(title=f\"$r={corr[0, 1]:.2f}$\", loc=\"upper left\")\n",
    "\n",
    "for idx, mod_x in enumerate(mods_x):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_x],\n",
    "        p_representations[mod_x],\n",
    "        metric=\"euclidean\"\n",
    "    )\n",
    "    x, y = get_values_from_dist_mat(x, mods_y, keep_diag=False)\n",
    "    corr = np.corrcoef(x, y)\n",
    "    mod_x_label = mod_x.capitalize()\n",
    "\n",
    "    # Sample a subset of the data for visualization\n",
    "    rnd_idx = np.random.choice(len(x), 15000, replace=False)\n",
    "    x, y = x[rnd_idx], y[rnd_idx]\n",
    "    \n",
    "    ax[1, idx].scatter(x, y, alpha=1, s=1)\n",
    "    ax[1, idx].set_xlabel(f\"Euclidean Distance ({mod_x}-{mod_x})\")\n",
    "    ax[1, idx].set_ylabel(\"Tanimoto Similarity\")\n",
    "    ax[1, idx].set_title(f\"Latent {mod_x_label} Euclidean Dist. vs Tanimoto Sim.\")\n",
    "    ax[1, idx].legend(title=f\"$r={corr[0, 1]:.2f}$\", loc=\"upper right\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"single_modality_vs_tanimoto.pdf\", dpi=300, format=\"pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotting different modalities of the same samples against each other\n",
    "mods = ['struct', 'morph', 'joint']\n",
    "\n",
    "fig, ax = plt.subplots(1, 3, figsize=(15, 5), squeeze=False)\n",
    "ctr = 0\n",
    "\n",
    "for idx in range(len(mods)):\n",
    "    for idy in range(idx+1, len(mods)):\n",
    "        x = get_pairwise_similarity(\n",
    "            p_representations[mods[idx]],\n",
    "            p_representations[mods[idx]],\n",
    "            metric=\"cosine\"\n",
    "        )\n",
    "        y = get_pairwise_similarity(\n",
    "            p_representations[mods[idy]],\n",
    "            p_representations[mods[idy]],\n",
    "            metric=\"cosine\"\n",
    "        )\n",
    "        x, y = get_values_from_dist_mat(x, y, keep_diag=False)\n",
    "        corr = np.corrcoef(x, y)\n",
    "\n",
    "        # Sample a subset of the data for visualization\n",
    "        rnd_idx = np.random.choice(len(x), 15000, replace=False)\n",
    "        x, y = x[rnd_idx], y[rnd_idx]\n",
    "        \n",
    "        ax_idx = 0\n",
    "\n",
    "        ax[ax_idx, ctr%3].scatter(x, y, alpha=1, s=1)\n",
    "        ax[ax_idx, ctr%3].set_xlabel(f\"Cosine Similarity ({mods[idx]}-{mods[idx]})\")\n",
    "        ax[ax_idx, ctr%3].set_ylabel(f\"Cosine Similarity ({mods[idy]}-{mods[idy]})\")\n",
    "        ax[ax_idx, ctr%3].set_title(f\"Latent {mods[idx].capitalize()} vs {mods[idy].capitalize()} Cosine Similarity\")\n",
    "        ax[ax_idx, ctr%3].legend(title=f\"$r={corr[0, 1]:.2f}$\", loc=\"upper left\")\n",
    "        ctr += 1\n",
    "        \n",
    "plt.tight_layout()\n",
    "plt.savefig(\"single_modality_vs_single_modality.pdf\", dpi=300, format=\"pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot cross-modality latent similarity versus raw data domain similarity (gene expression & Tanimoto sim)\n",
    "mods_x = ['struct', 'ge', 'joint']\n",
    "ge_sim = get_pairwise_similarity(ge, ge, metric=\"cosine\")\n",
    "\n",
    "fig, ax = plt.subplots(3, 3, figsize=(15, 12))\n",
    "\n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    x, y = get_values_from_dist_mat(x, tanimoto_sim, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    ax[0, idx].scatter(x, y, alpha=0.5, s=.1)\n",
    "    ax[0, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[0, idx].set_ylabel(\"Tanimoto Similarity\")\n",
    "    ax[0, idx].set_title(f\"({mod_a}-{mod_b}) vs Tanimoto (r={corr[0, 1]:.2f})\")\n",
    "    \n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    y = ge_sim\n",
    "    x, y = get_values_from_dist_mat(x, y, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    ax[1, idx].scatter(x, y, alpha=0.5, s=.1)\n",
    "    ax[1, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[1, idx].set_ylabel(\"Cosine Similarity (raw gene exp)\")\n",
    "    ax[1, idx].set_title(f\"({mod_a}-{mod_b}) vs raw GE (r={corr[0, 1]:.2f})\")\n",
    "    \n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    y = ge_sim\n",
    "    x, y = get_values_from_dist_mat(x, y, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    ax[2, idx].scatter(x, y, alpha=0.5, s=.1)\n",
    "    ax[2, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[2, idx].set_ylabel(\"Euclidean Distance (raw gene exp)\")\n",
    "    ax[2, idx].set_title(f\"({mod_a}-{mod_b}) vs raw GE (r={corr[0, 1]:.2f})\")\n",
    "        \n",
    "plt.tight_layout()\n",
    "plt.savefig(\"cross_modality_vs_raw_sim.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot cross-modality latent similarity versus PCA morph data\n",
    "mods_x = ['struct', 'morph', 'joint']\n",
    "morph_pca_sim_cos = get_pairwise_similarity(morph_pca, morph_pca, metric=\"cosine\")\n",
    "morph_pca_sim_euc = get_pairwise_similarity(morph_pca, morph_pca, metric=\"euclidean\")\n",
    "\n",
    "fig, ax = plt.subplots(2, 3, figsize=(15, 10))\n",
    "\n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    x, y = get_values_from_dist_mat(x, tanimoto_sim, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    # Sample a subset of the data for visualization\n",
    "    rnd_idx = np.random.choice(len(x), 15000, replace=False)\n",
    "    x, y = x[rnd_idx], y[rnd_idx]\n",
    "\n",
    "    ax[0, idx].scatter(x, y, alpha=1, s=1)\n",
    "    ax[0, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[0, idx].set_ylabel(\"Tanimoto Similarity\")\n",
    "    ax[0, idx].set_title(f\"Cross-Modal Latent Cosine Sim. ({mod_a}-{mod_b})\\nvs Tanimoto Sim.\")\n",
    "    ax[0, idx].legend(title=f\"$r={corr[0, 1]:.2f}$\", loc=\"upper left\")\n",
    "    \n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    y = morph_pca_sim_cos\n",
    "    x, y = get_values_from_dist_mat(x, y, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    # Sample a subset of the data for visualization\n",
    "    rnd_idx = np.random.choice(len(x), 15000, replace=False)\n",
    "    x, y = x[rnd_idx], y[rnd_idx]\n",
    "\n",
    "    ax[1, idx].scatter(x, y, alpha=1, s=1)\n",
    "    ax[1, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[1, idx].set_ylabel(\"Cosine Similarity (morph-morph PCA)\")\n",
    "    ax[1, idx].set_title(f\"Cross-Modal Latent Cosine Sim. ({mod_a}-{mod_b})\\nvs Morph PCA Cosine Sim.\")\n",
    "    ax[1, idx].legend(title=f\"$r={corr[0, 1]:.2f}$\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"cross_modality_vs_pca_morph.pdf\", dpi=300, format=\"pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "morph_pca_sim_cos = get_pairwise_similarity(morph_pca, morph_pca, metric=\"cosine\")\n",
    "morph_pca_sim_euc = get_pairwise_similarity(morph_pca, morph_pca, metric=\"euclidean\")\n",
    "\n",
    "fig, ax = plt.subplots(1, 2, figsize=(10, 5))\n",
    "\n",
    "for idx, (mod_a, mod_b) in enumerate([('struct', 'morph'), ('struct', 'joint')]):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    y = morph_pca_sim_cos\n",
    "    x, y = get_values_from_dist_mat(x, y, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    # Keep only a subset of the data for visualization\n",
    "    rnd_idx = np.random.choice(len(x), 15000, replace=False)\n",
    "    x, y = x[rnd_idx], y[rnd_idx]\n",
    "\n",
    "    ax[idx].scatter(x, y, alpha=1, s=1)\n",
    "    ax[idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[idx].set_ylabel(\"Cosine Similarity (morph-morph PCA)\")\n",
    "    ax[idx].set_title(f\"Cross-Modal Latent Cosine Sim. ({mod_a.capitalize()}-{mod_b.capitalize()})\\nvs Morph PCA Cosine Sim.\")\n",
    "    ax[idx].legend(title=f\"$r={corr[0, 1]:.2f}$\", loc=\"upper left\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"struct&joint_vs_pca_morph.pdf\", dpi=300, format=\"pdf\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot cross-modality latent similarity versus raw morph domain distance\n",
    "mods_x = ['struct', 'morph', 'joint']\n",
    "morph_sim_cos = get_pairwise_similarity(morph, morph, metric=\"cosine\")\n",
    "morph_sim_euc = get_pairwise_similarity(morph, morph, metric=\"euclidean\")\n",
    "\n",
    "fig, ax = plt.subplots(2, 3, figsize=(15, 8))\n",
    "\n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    y = morph_sim_cos\n",
    "    x, y = get_values_from_dist_mat(x, y, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    ax[0, idx].scatter(x, y, alpha=0.5, s=.1)\n",
    "    ax[0, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[0, idx].set_ylabel(f\"Cosine Similarity (raw morph features)\")\n",
    "    ax[0, idx].set_title(f\"({mod_a}-{mod_b}) vs raw morph (r={corr[0, 1]:.2f})\")\n",
    "    \n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    y = morph_sim_euc\n",
    "    x, y = get_values_from_dist_mat(x, y, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    ax[1, idx].scatter(x, y, alpha=0.5, s=.1)\n",
    "    ax[1, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[1, idx].set_ylabel(f\"Euclidean Distance (raw morph features)\")\n",
    "    ax[1, idx].set_title(f\"({mod_a}-{mod_b}) vs raw morph (r={corr[0, 1]:.2f})\")\n",
    "        \n",
    "plt.tight_layout()\n",
    "plt.savefig(\"cross_modality_vs_raw_morph.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot cross-modality latent similarity versus PCA gene expression\n",
    "mods_x = ['struct', 'ge', 'joint']\n",
    "ge_pca_sim_cos = get_pairwise_similarity(ge_pca, ge_pca, metric=\"cosine\")\n",
    "ge_pca_sim_euc = get_pairwise_similarity(ge_pca, ge_pca, metric=\"euclidean\")\n",
    "\n",
    "fig, ax = plt.subplots(2, 3, figsize=(15, 8))\n",
    "    \n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    y = ge_pca_sim_cos\n",
    "    x, y = get_values_from_dist_mat(x, y, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    ax[0, idx].scatter(x, y, alpha=0.5, s=.1)\n",
    "    ax[0, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[0, idx].set_ylabel(f\"Cosine Similarity (PCA gene exp, n={N_genes})\")\n",
    "    ax[0, idx].set_title(f\"({mod_a}-{mod_b}) vs pca GE (r={corr[0, 1]:.2f})\")\n",
    "    \n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    y = ge_pca_sim_euc\n",
    "    x, y = get_values_from_dist_mat(x, y, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    ax[1, idx].scatter(x, y, alpha=0.5, s=.1)\n",
    "    ax[1, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[1, idx].set_ylabel(f\"Euclidean Distance (PCA gene exp, n={N_genes})\")\n",
    "    ax[1, idx].set_title(f\"({mod_a}-{mod_b}) vs pca GE (r={corr[0, 1]:.2f})\")\n",
    "        \n",
    "plt.tight_layout()\n",
    "plt.savefig(\"cross_modality_vs_pca_ge.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run same analysis as above, but with the top-k=81 most differentially expressed genes obtained from file\n",
    "mods_x = ['struct', 'ge', 'joint']\n",
    "filtered_ge_sim_cos = get_pairwise_similarity(filtered_ge, filtered_ge, metric=\"cosine\")\n",
    "filtered_ge_sim_euc = get_pairwise_similarity(filtered_ge, filtered_ge, metric=\"euclidean\")\n",
    "# filtered_ge_sim_mse = get_pairwise_similarity(filtered_ge, filtered_ge, metric=\"mse\")\n",
    "\n",
    "fig, ax = plt.subplots(2, 3, figsize=(15, 8))\n",
    "    \n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    y = filtered_ge_sim_cos\n",
    "    x, y = get_values_from_dist_mat(x, y, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    ax[0, idx].scatter(x, y, alpha=0.5, s=.1)\n",
    "    ax[0, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[0, idx].set_ylabel(f\"Cosine Similarity (DE gene exp, n=81)\")\n",
    "    ax[0, idx].set_title(f\"({mod_a}-{mod_b}) vs filtered GE (r={corr[0, 1]:.2f})\")\n",
    "    \n",
    "for idx, (mod_a, mod_b) in enumerate(itertools.combinations(mods_x, 2)):\n",
    "    x = get_pairwise_similarity(\n",
    "        p_representations[mod_a],\n",
    "        p_representations[mod_b],\n",
    "        metric=\"cosine\"\n",
    "    )\n",
    "    y = filtered_ge_sim_euc\n",
    "    x, y = get_values_from_dist_mat(x, y, keep_diag=True)\n",
    "    corr = np.corrcoef(x, y)\n",
    "\n",
    "    ax[1, idx].scatter(x, y, alpha=0.5, s=.1)\n",
    "    ax[1, idx].set_xlabel(f\"Cosine Similarity ({mod_a}-{mod_b})\")\n",
    "    ax[1, idx].set_ylabel(f\"Euclidean Distance (DE gene exp, n=81)\")\n",
    "    ax[1, idx].set_title(f\"({mod_a}-{mod_b}) vs filtered GE (r={corr[0, 1]:.2f})\")\n",
    "        \n",
    "plt.tight_layout()\n",
    "plt.savefig(\"cross_modality_vs_filtered_ge_top_81.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sanity Checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn import preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_ge = np.load(\"adata_scale_filt.npy\")\n",
    "scaled_filtered_ge = preprocessing.normalize(filtered_ge, axis=1)\n",
    "filtered_test_ge = filtered_ge[test_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "loader = make_eval_data_loader(datamodule.dataset)\n",
    "mods = unroll_dataloader(loader, mods=['ge', 'morph'])\n",
    "ge = mods['ge']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check that there is overlap in the gene expression columns between the DE gene subset and the original\n",
    "print(filtered_ge.shape, ge.shape)\n",
    "\n",
    "de_idx = []\n",
    "\n",
    "for col_1 in tqdm(filtered_ge.T):\n",
    "    for idx, col_2 in enumerate(ge.T):\n",
    "        if np.array_equal(col_1, col_2):\n",
    "            de_idx.append(idx)\n",
    "            \n",
    "len(de_idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute mean and std deviation of the raw gene expression\n",
    "u = np.mean(ge)\n",
    "v = np.std(ge)\n",
    "\n",
    "nu = np.mean(filtered_ge)\n",
    "nv = np.std(filtered_ge)\n",
    "\n",
    "snu = np.mean(scaled_filtered_ge)\n",
    "snv = np.std(scaled_filtered_ge)\n",
    "\n",
    "print(u, nu)\n",
    "print(v, nv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(np.linalg.norm(ge[0]))\n",
    "print(np.linalg.norm(filtered_ge[0]))\n",
    "print(np.linalg.norm(scaled_filtered_ge[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot histogram of pairwise distances in the raw gene expression space versus the filtered gene expression space\n",
    "ge_dist = get_pairwise_similarity(ge, ge, metric=\"euclidean\")\n",
    "filtered_ge_dist = get_pairwise_similarity(filtered_ge, filtered_ge, metric=\"euclidean\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y = get_values_from_dist_mat(ge_dist, filtered_ge_dist)\n",
    "x.shape, y.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 2, figsize=(10, 3))\n",
    "\n",
    "ax[0].hist(x, density=True, bins=40)\n",
    "ax[0].set_ylabel('Density')\n",
    "ax[0].set_xlabel('Eucl. dist. between pairs')\n",
    "ax[0].set_title('Pairwise dist. in raw gene expr. data')\n",
    "\n",
    "ax[1].hist(y, density=True, bins=40)\n",
    "ax[1].set_ylabel('Density')\n",
    "ax[1].set_xlabel('Eucl. dist. between pairs')\n",
    "ax[1].set_title('Pairwise dist. in filtered gene expr. data')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(np.sort(ge[0]))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(np.sort(ge[1]))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mmc_sim",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
